### PR TITLE
The static checks of the TransformerEncoder should consider ...

### PR DESCRIPTION
**Pull request by gaolexx**
**Transformer support num_encoder_layers=0 as PyTorch 1.6**

The static checks of the TransformerEncoder should consider num_layers to avoid IndexError.

The `Transformer` can be used to implement all the capabilities of the `TransformerEncoder` and `TransformerDecoder`. Therefore it is also possible to set the `Transformer` initialization parameter `num_encoder_layers` to `0`. In this case the current static check (`first_layer = self.layers[0]`) will throw an exception `IndexError: index 0 is out of range`. This pr is to fix this bug of out-of-bounds array access.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg

### CODE REVIEW

**Comment by gao-lex** [comment]
cc @mikekgfb

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -261,59 +262,60 @@ def forward(
...
-                first_layer.norm1.weight,
-                first_layer.norm1.bias,
-                first_layer.norm2.weight,
-                first_layer.norm2.bias,
-                first_layer.linear1.weight,
-                first_layer.linear1.bias,
-                first_layer.linear2.weight,
-                first_layer.linear2.bias,
-            )
+        if self.num_layers > 0:
```
Comment (mikekgfb) [inline torch/nn/modules/transformer.py:266]: We know that self.num_layers == 0 in __init__ already, so we can set `use_nested_tensors=False` and prevent all references to first_layer being used. Replace 1 reference to first_layer in "if/elif" chain with self.layer[0], and move `first_layer = self.layers[0]` down just above tensor_args?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -196,27 +196,28 @@ def __init__(self, encoder_layer, num_layers, norm=None, enable_nested_tensor=Tr
...
-            why_not_sparsity_fast_path = f" {str_first_layer}.activation_relu_or_gelu was not True"
-        elif not (first_layer.norm1.eps == first_layer.norm2.eps) :
-            why_not_sparsity_fast_path = f"{str_first_layer}.norm1.eps was not equal to {str_first_layer}.norm2.eps"
-        elif first_layer.self_attn.num_heads % 2 == 1:
-            why_not_sparsity_fast_path = "num_head is odd"
-
-        if enable_nested_tensor and why_not_sparsity_fast_path:
-            warnings.warn(f"enable_nested_tensor is True, but self.use_nested_tensor is False because {why_not_sparsity_fast_path}")
-            self.use_nested_tensor = False
+        if self.num_layers > 0:
```
Comment (mikekgfb) [inline torch/nn/modules/transformer.py:199]: replace all references to `first_layer` in `__init__` with references to `encoder_layer` and make num_layers another reason why to set `use_nested_tensor=False`

**Comment by mikaylagawarecki** [comment]
This is fixed by #104058

**Pull request closed without merging**